### PR TITLE
[MAINTENANCE] Adding type hints to DataAssistant implementations

### PR DESCRIPTION
**Pull request by alexsherstinsky**
Please annotate your PR title to describe what the PR does, then give a brief bulleted description of your PR below. PR titles should begin with [BUGFIX], [FEATURE],  [DOCS], or [MAINTENANCE]. If a new feature introduces breaking changes for the Great Expectations API or configuration files, please also add [BREAKING]. You can read about the tags in our [contributor checklist](https://docs.greatexpectations.io/docs/contributing/contributing_checklist).

Changes proposed in this pull request:
-
-
-

After submitting your PR, CI checks will run and @cla-bot will check for your CLA signature.

For a PR with nontrivial changes, we review with both design-centric and code-centric lenses.

In a design review, we aim to ensure that the PR is consistent with our relationship to the open source community, with our software architecture and abstractions, and with our users' needs and expectations. That review often starts well before a PR, for example in github issues or slack, so please link to relevant conversations in notes below to help reviewers understand and approve your PR more quickly (e.g. `closes #123`).

Previous Design Review notes:
-
-
-

### Definition of Done
Please delete options that are not relevant.

- [x] My code follows the Great Expectations [style guide](https://docs.greatexpectations.io/docs/contributing/style_guides/code_style)
- [x] I have performed a [self-review](https://docs.greatexpectations.io/docs/contributing/contributing_checklist) of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] I have added [unit tests](https://docs.greatexpectations.io/docs/contributing/contributing_test#writing-unit-and-integration-tests) where applicable and made sure that new and existing tests are passing.
- [x] I have run any local integration tests and made sure that nothing is broken.

Thank you for submitting!

### CODE REVIEW

File: great_expectations/rule_based_profiler/data_assistant/volume_data_assistant.py [hunk 1]
```diff
@@ -8,6 +8,7 @@
 )
 from great_expectations.rule_based_profiler.domain_builder import (
     CategoricalColumnDomainBuilder,
+    DomainBuilder,
```
Comment (cdkini) [inline great_expectations/rule_based_profiler/data_assistant/volume_data_assistant.py]: If only importing for the purposes of type hinting, please use the following pattern:

```python
from typing import TYPE_CHECKING

if TYPE_CHECKING:
    from great_expectations...
```

File: great_expectations/rule_based_profiler/data_assistant/onboarding_data_assistant.py [hunk 2]
```diff
@@ -614,7 +615,7 @@ def _build_datetime_columns_rule() -> Rule:
 
         # Step-1: Instantiate "ColumnDomainBuilder" for selecting proper datetime columns (not "datetime-looking" text).
 
-        datetime_column_type_domain_builder = ColumnDomainBuilder(
+        datetime_column_type_domain_builder: DomainBuilder = ColumnDomainBuilder(
```
Comment (cdkini) [inline great_expectations/rule_based_profiler/data_assistant/onboarding_data_assistant.py:620]: I don't think we need type hints here - the type is inferred from the right-hand side of the expression and the relation to the parent is implicit

**Review by cdkini — CHANGES_REQUESTED** [review]

File: great_expectations/rule_based_profiler/data_assistant/onboarding_data_assistant.py (hunk 2 above)
Comment (NathanFarmer) [reply great_expectations/rule_based_profiler/data_assistant/onboarding_data_assistant.py:620]: Agreed

File: great_expectations/rule_based_profiler/data_assistant/volume_data_assistant.py (hunk 1 above)
Comment (NathanFarmer) [reply great_expectations/rule_based_profiler/data_assistant/volume_data_assistant.py]: Agreed

File: great_expectations/rule_based_profiler/data_assistant/onboarding_data_assistant.py (hunk 2 above)
Comment (alexsherstinsky) [reply great_expectations/rule_based_profiler/data_assistant/onboarding_data_assistant.py:620]: @cdkini This is to emphasize the polymorphism.  Thanks.

**Review by cdkini — APPROVED** [review]

**Comment by alexsherstinsky** [comment]
@cla-bot check

**Review by Kilo59 — CHANGES_REQUESTED** [review]
~I don't think these modules are being type-checked. Can we fix that first?~

https://github.com/great-expectations/great_expectations/blob/develop/pyproject.toml#L104

Nevermind. Was looking at the results module.

**Review by Kilo59 — APPROVED** [review]

**Pull request merged**